src/drivers/uavcan/node_info.cpp

```diff
@@ -0,0 +1,318 @@
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "node_info.hpp"
+
+#include <px4_platform_common/log.h>
+#include <px4_platform_common/px4_config.h>
+#include <drivers/drv_hrt.h>
+#include <cstring>
+
+using namespace time_literals;
+
+NodeInfoPublisher::NodeInfoPublisher(uavcan::INode &node, uavcan::NodeInfoRetriever &node_info_retriever)
+	: TimerBase(node), _node_info_retriever(node_info_retriever)
+{
+	_node_info_retriever.addListener(this);
+}
+
+NodeInfoPublisher::~NodeInfoPublisher()
+{
+	_node_info_retriever.removeListener(this);
+	delete[] _device_informations;
+}
+
+void NodeInfoPublisher::handleNodeInfoRetrieved(uavcan::NodeID node_id,
+		const uavcan::protocol::GetNodeInfo_::Response &node_info)
+{
+	const NodeInfo info(node_id, node_info);
+	registerDevice(info.node_id.get(), &info, 0, DeviceCapability::NONE);
+
+	startTimerIfNotRunning();
+}
+
+void NodeInfoPublisher::handleNodeInfoUnavailable(uavcan::NodeID node_id)
+{
+}
+
+void NodeInfoPublisher::handleTimerEvent(const uavcan::TimerEvent &event)
+{
+	// Publish device information using round-robin approach
+	publishDeviceInformationPeriodic();
+}
+
+void NodeInfoPublisher::startTimerIfNotRunning()
+{
+	if (!TimerBase::isRunning()) {
+		TimerBase::startPeriodic(uavcan::MonotonicDuration::fromMSec(DEVICE_INFO_PUBLISH_INTERVAL_MS));
+	}
+}
+
+void NodeInfoPublisher::registerDevice(uint8_t node_id, const NodeInfo *info, uint32_t device_id,
+				       DeviceCapability capability)
+{
+	const bool is_registering_info = (info != nullptr);
```

Comment:
hmm I'm not sure I like the overloading but it might be better than the alternative